pkg/query/storeset.go

```diff
@@ -0,0 +1,258 @@
+package query
+
+import (
+	"context"
+	"math"
+	"sync"
+	"time"
+
+	"fmt"
+
+	"github.com/go-kit/kit/log"
+	"github.com/go-kit/kit/log/level"
+	"github.com/grpc-ecosystem/go-grpc-middleware"
+	"github.com/grpc-ecosystem/go-grpc-prometheus"
+	"github.com/improbable-eng/thanos/pkg/store"
+	"github.com/improbable-eng/thanos/pkg/store/storepb"
+	"github.com/improbable-eng/thanos/pkg/tracing"
+	"github.com/opentracing/opentracing-go"
+	"github.com/pkg/errors"
+	"github.com/prometheus/client_golang/prometheus"
+	"golang.org/x/sync/errgroup"
+	"google.golang.org/grpc"
+)
+
+// StoreSet maintains a set of active stores. It is backed by a peer's view of the cluster
+// and a list of static store addresses.
+type StoreSet struct {
+	logger           log.Logger
+	peerAddrs        func() []string
+	staticAddrs      []string
+	dialOpts         []grpc.DialOption
+	gRPCRetryTimeout time.Duration
+
+	mtx                  sync.RWMutex
+	stores               map[string]*storeRef
+	storeNodeConnections prometheus.Gauge
+}
+
+// NewStoreSet returns a new set of stores from cluster peers and statically configured ones.
+func NewStoreSet(
+	logger log.Logger,
+	reg *prometheus.Registry,
+	tracer opentracing.Tracer,
+	peerAddrs func() []string,
+	staticAddrs []string,
```

Comment:
Do we need those both? From a first glance it doesn't seem we distinguish how we treat peers vs static. Can the caller just return the union of both in the getter func?

Reply:
Yea.. we use that for better logging. But TBH not sure if we actually need that.

Reply:
also we will lose control of choosing the better "one" in case of duplicates... I wonder we could catch that early on?